internal/pkg/agent/application/coordinator/diagnostics_test.go

```diff
@@ -0,0 +1,397 @@
+	helper.runtimeManager.EXPECT().SubscribeAll(mock.Anything).Return(subscriptionAll)
+	helper.runtimeManager.EXPECT().Update(mock.AnythingOfType("[]component.Component")).Return(nil)
+
+	sut := helper.coordinator
+	coordinatorWg := new(sync.WaitGroup)
+	defer func() {
+		cancelFunc()
+		// wait till the coordinator exits to avoid a panic
+		// when logging after the test goroutine exited
+		coordinatorWg.Wait()
+	}()
+
+	coordinatorWg.Add(1)
+	go func() {
+		defer coordinatorWg.Done()
+		coordErr := sut.Run(ctx)
+		assert.ErrorIs(t, coordErr, context.Canceled, "Coordinator exited with unexpected error")
+	}()
+
+	//Provide vars
+	processors := transpiler.Processors{
+		{
+			"add_fields": map[string]interface{}{
+				"dynamic": "added",
+			},
+		},
+	}
+	fetchContextProvider := mocks.NewFetchContextProvider(t)
+	fetchContextProviders := mapstr.M{
+		"kubernetes_secrets": fetchContextProvider,
+	}
+	vars, err := transpiler.NewVarsWithProcessors(
+		"id",
+		map[string]interface{}{
+			"host": map[string]interface{}{"platform": "linux"},
+			"dynamic": map[string]interface{}{
+				"key1": "dynamic1",
+				"list": []string{
+					"array1",
+					"array2",
+				},
+				"dict": map[string]string{
+					"key1": "value1",
+					"key2": "value2",
+				},
+			},
+		},
+		"dynamic",
+		processors,
+		fetchContextProviders)
+	require.NoError(t, err)
+	mustWriteToChannelBeforeTimeout(t, []*transpiler.Vars{vars}, helper.varsChannel, 100*time.Millisecond)
+
+	// Inject initial configuration - after starting coordinator
+	configBytes, err := os.ReadFile("./testdata/simple_config/elastic-agent.yml")
+	require.NoError(t, err)
+
+	initialConf := config.MustNewConfigFrom(configBytes)
+
+	initialConfChange := mocks.NewConfigChange(t)
+	initialConfChange.EXPECT().Config().Return(initialConf)
+	initialConfChange.EXPECT().Ack().Return(nil).Times(1)
+	mustWriteToChannelBeforeTimeout[coordinator.ConfigChange](t, initialConfChange, helper.configChangeChannel, 100*time.Millisecond)
+
+	assert.Eventually(t, func() bool { return sut.State().State == cproto.State_HEALTHY }, 1*time.Second, 50*time.Millisecond)
+	t.Logf("Agent state: %s", sut.State().State)
+
+	// Runtime component state
+	componentState := runtime.ComponentComponentState{
+		Component: component.Component{
+			ID: "mock_component_1",
+			Units: []component.Unit{
+				{
+					ID:       "mock_input_unit",
+					Type:     client.UnitTypeInput,
+					LogLevel: client.UnitLogLevelInfo,
+					Config:   &proto.UnitExpectedConfig{},
+				},
+			},
+		},
+		State: runtime.ComponentState{
+			State: client.UnitStateHealthy,
+			VersionInfo: runtime.ComponentVersionInfo{
+				Name:    "shiny mock component",
+				Version: "latest, obvs :D",
+			},
+		},
+	}
+	mustWriteToChannelBeforeTimeout(t, componentState, componentsUpdateChannel, 100*time.Millisecond)
+
+	// FIXME there's no way to know if the coordinator processed the runtime component states, wait and hope for the best
+	time.Sleep(50 * time.Millisecond)
```

Comment:
as a consequence of using `coordinator_test` package to avoid import loops from the mocks generated by coordinator interfaces that live in a different package (more on a dedicated comment) we cannot access in this test anything that is not exported by  `coordinator` package